Notebook based on [Design](https://nightly.docs.classiq.io/latest/classiq_101/classiq_concepts/design/)

# Concrete Example
First, understand Qmod through an example.

The task is to design a quantum algorithm that coherently computes the arithmetic operation $y=x^2+1$, for a quantum variable $\Ket{x}$ taht is a superposition of all the numbers between 0 and 7: $\Ket{x} = \frac{1}{\sqrt{8}}(\Ket{0}+\Ket{1}+ ... + \Ket{7})$. The expected output is
$$\Ket{x}\Ket{y} = \Ket{x}\Ket{x^2+1} = \frac{1}{\sqrt{8}}\sum^{7}_{i=0}\Ket{i}\Ket{i^2+1}$$, where $\Ket{x}$ is entangled to $\Ket{y}$.
This may sound complicated, but the following few lines of code in Qmod create the desired algorithm with Classiq.

In [25]:
from classiq import Output, QNum, allocate, hadamard_transform, qfunc

@qfunc
def main(x: Output[QNum],y: Output[QNum]):
    allocate(4, x)
    hadamard_transform(x) # creates a uniform superposition
    y |= x**2 + 1

In [3]:
from classiq import create_model, show, synthesize

quantum_program = synthesize(create_model(main))
show(quantum_program)


Opening: https://platform.classiq.io/circuit/094c468a-f344-4e96-b799-2c69dda4aca9?version=0.42.2


In [5]:
from classiq import execute
job = execute(quantum_program)
results = job.result()[0].value.parsed_counts
print(results)

[{'x': 1.0, 'y': 2.0}: 141, {'x': 10.0, 'y': 101.0}: 141, {'x': 5.0, 'y': 26.0}: 135, {'x': 3.0, 'y': 10.0}: 135, {'x': 6.0, 'y': 37.0}: 135, {'x': 0.0, 'y': 1.0}: 133, {'x': 11.0, 'y': 122.0}: 131, {'x': 13.0, 'y': 170.0}: 129, {'x': 12.0, 'y': 145.0}: 129, {'x': 9.0, 'y': 82.0}: 128, {'x': 2.0, 'y': 5.0}: 124, {'x': 14.0, 'y': 197.0}: 123, {'x': 15.0, 'y': 226.0}: 117, {'x': 8.0, 'y': 65.0}: 117, {'x': 4.0, 'y': 17.0}: 117, {'x': 7.0, 'y': 50.0}: 113]


The above code is a quantum model written in Qmod. A model is composed of quantum functions with at least one quantum function called main. Define quantum functions in Python using regular Python functions decorated with @qfunc, and in the native Qmod syntax define them with the qfunc keyword.

Quantum functions manipulate quantum objects that are represented using quantum variables. Declare and initialize each variable before it is used.

### Additional exercise 1:
**Task**: Write within the IDE with the native syntax a quantum algorithm that coherently calculates y= 2*x + 4*z +2, where x and z are quantum numbers represented with 5 qubits each, and they are initialized in a superposition of all possible values.

In [22]:
from classiq import Output, QNum, allocate, hadamard_transform, qfunc

@qfunc
def main(x: Output[QNum], z: Output[QNum], y: Output[QNum]):
    allocate(5, x)
    allocate(5, z)
    hadamard_transform(x) # creates a uniform
    hadamard_transform(z) # creates a uniform# superposition
    y |= 2*x + 4*z + 2

In [23]:
from classiq import create_model, show, synthesize

quantum_program = synthesize(create_model(main))
show(quantum_program)

Opening: https://platform.classiq.io/circuit/49de6796-13b4-4138-ad0d-0e38d367681a?version=0.42.2


In [24]:
from classiq import execute
job = execute(quantum_program)
results = job.result()[0].value.parsed_counts
print(results)

[{'x': 23.0, 'z': 1.0, 'y': 52.0}: 10, {'x': 18.0, 'z': 20.0, 'y': 118.0}: 8, {'x': 0.0, 'z': 10.0, 'y': 42.0}: 7, {'x': 20.0, 'z': 9.0, 'y': 78.0}: 6, {'x': 22.0, 'z': 21.0, 'y': 130.0}: 6, {'x': 14.0, 'z': 20.0, 'y': 110.0}: 6, {'x': 14.0, 'z': 24.0, 'y': 126.0}: 6, {'x': 18.0, 'z': 4.0, 'y': 54.0}: 6, {'x': 12.0, 'z': 26.0, 'y': 130.0}: 6, {'x': 1.0, 'z': 27.0, 'y': 112.0}: 6, {'x': 13.0, 'z': 30.0, 'y': 148.0}: 6, {'x': 30.0, 'z': 16.0, 'y': 126.0}: 6, {'x': 22.0, 'z': 11.0, 'y': 90.0}: 6, {'x': 27.0, 'z': 31.0, 'y': 180.0}: 6, {'x': 5.0, 'z': 28.0, 'y': 124.0}: 6, {'x': 26.0, 'z': 18.0, 'y': 126.0}: 6, {'x': 5.0, 'z': 22.0, 'y': 100.0}: 6, {'x': 3.0, 'z': 31.0, 'y': 132.0}: 6, {'x': 11.0, 'z': 13.0, 'y': 76.0}: 6, {'x': 26.0, 'z': 31.0, 'y': 178.0}: 6, {'x': 11.0, 'z': 25.0, 'y': 124.0}: 6, {'x': 28.0, 'z': 19.0, 'y': 134.0}: 6, {'x': 10.0, 'z': 6.0, 'y': 46.0}: 6, {'x': 19.0, 'z': 11.0, 'y': 84.0}: 5, {'x': 15.0, 'z': 5.0, 'y': 52.0}: 5, {'x': 18.0, 'z': 22.0, 'y': 126.0}: 5, {'x

In [26]:
from classiq import write_qmod
write_qmod(create_model(main), "design- additional exercise")

1.  Include a qfunc decorator in Python or keyword in the native syntax
You can see the qfunc decorator (@qfunc) in the Python implementation and the corresponding keyword in the native implementation. This indicates to Qmod that we are dealing with quantum functions that manipulate quantum objects.
2. Include a main function
Every quantum algorithm written in Qmod must have a main function. From this main function, create the quantum program using the synthesize(create_model(main)) command in Python or by clicking Synthesis in the IDE.
3.  Declare the arguments of the main function as outputs
In the above example, the two arguments of the main function—x and y—are both quantum variables. Their type is QNum(qnum in the native syntax), which announces a quantum number (see Quantum Variables). In addition to the type, declare the variables of the main function as output, indicating that these quantum variables are not initialized outside the scope of the function.
4.  Declare each quantum variable before using it
5.  Initialize each quantum variable

After declaring a quantum variable, initialize it using one of several ways. Two are shown in the example above:

x is initialized with the allocate operation.
y i initialized with the |= (= in native) numeric assignment.

In [6]:
from classiq import write_qmod
write_qmod(create_model(main), "design")

# Design - Quantum Variables and Functions
In Qmod there are three types of quantum objects which are represented by quantum variables:

1.**Qubit** (quantum bit)
* Written as `QBit` in the Python SDK
* Written as `qbit` in the native syntax

2.**Qubit Array**:
A sequence of qubits, should be thought of as a quantum register with definite number of qubits.
* Written as `QArray` in the Python SDK
* Written as `qbit[]` in the native syntax

3.**Quantum Number**:
A qubit array (quantum register) that represents numbers. The numbers can be positive and negative integers, and it can be fixed point numbers (e.g.$-5,25$). The quantum number object has a definite number of qubits (as every qubit array), and it has 2 properties for its numeric representation: if the number is signed or not (only positive or it represents negative numbers as well) and how many fractional digits it has (where is the decimal point).
* Written as `QNum` in the Python SDK
* Written as `qnum` in the native syntax

### Example
Our task is to design a quantum algorithm that flips the most significant bit (MSB) of a quantum number and to verify this.
How to approach that? We will have a function called flip_msb that will receive a Qubit Array, and we will flip its MSB. This function will be called from the main function with some Quantum Number, and then we will verify we receive the correct number using a Qubit indicator.
The flip_msb function receives a QArray variable named reg (shortcut for register), and it flips its MSB using the X gate. This gate operates on the qubit at the last position of the Qubit Array reg where the counting starts from. One can see the property len of the register reg is used as part of the function.

In [1]:
from classiq import QArray, X, qfunc

@qfunc
def flip_msb(reg: QArray):
    X(reg[reg.len - 1])

An important thing to note is that the variable reg is not initialized within the scope of the function flip_msb. This can be seen by the lack of the output declaration in the scope of the function, and it means that reg is initialized before it is called in the function.
Now to the main function.
We have only one variable as an argument for the function which is the indicator if we manage to flip the MSB or not. This is a Qubit variable named indicator.
Within the function itself we declare and initialize a Quantum Number named x with 4 qubits. A general initialziation with `allocate` initializes the Quantum Number to the state $\Ket{0} = \Ket{0000}$ (this is true for Qubits and Qubit Arrays as well). The function flip_msb acts on the Quantum Number x in order to flip its MSB and to create the state $\Ket{1000} = \Ket{8}$
.


In [3]:
from classiq import Output, QBit, QNum, allocate

@qfunc
def main(indicator: Output[QBit]):

    x = QNum("x")
    allocate(4,x)
    flip_msb(x)

    indicator |= x == 8

Finally, we initialize the indicator qubit with the numeric assignment of the expression $x == 8$. That is, if indeed we manage to flip the MSB of x and to transform it to the state $|8\rangle$ then the state od the qubit indicator will be equal $\Ket{1}$ otherwise it will equal $|0\rangle$ (i.e. indicating if we manage to complete the flip_msb operation or not).

Note:
**`flip_msb` receives a Quantum Number or a Quantum Array?**
One might notice that the function `flip_msb` is called with a Quantum Number `x` from the `main` function, whilst it is declared with a quantum array named `reg`. This demonstrates that quantum numbers can be cast to quantum arrays if they have the same number of qubits. In `main` we are intrested in the manupulation of an arithmetic quantum object - a quantum number, whilst in the function `flip_msb` we are just interested in treating the quantum number as a qubit array, without its numeric description, and to apply a specific gate on one of its qubits.

That's it! Now we want to check if indeed we managed to flip the MSB. The desired quantum program can be synthesized and viewed:

In [4]:
from classiq import create_model, show, synthesize

quantum_program = synthesize(create_model(main))
show(quantum_program)

Opening: https://platform.classiq.io/circuit/8aae5411-0d44-4fea-baf9-3cf60c184ad4?version=0.42.2


And then the quantum program can be executed to receive the results:

In [5]:
from classiq import execute

job = execute(quantum_program)
results = job.result()[0].value.parsed_counts
print(results)

[{'indicator': 1.0}: 2048]


Task: Create another quantum function flip_lsb that flips the least significant bit.

In [27]:
from classiq import Output, QBit, QNum, allocate
from classiq import QArray, X, qfunc

@qfunc
def flip_lsb(reg: QArray):
    X(reg[0])

@qfunc
def main(indicator: Output[QBit]):

    x = QNum("x")
    allocate(4,x)
    flip_msb(x)

    indicator |= x == 8

In [28]:
from classiq import create_model, show, synthesize

quantum_program = synthesize(create_model(main))
show(quantum_program)

Opening: https://platform.classiq.io/circuit/6cbbe6fd-8dc3-4877-a310-f9dc277e12d0?version=0.42.2


In [29]:
from classiq import execute

job = execute(quantum_program)
results = job.result()[0].value.parsed_counts
print(results)

[{'indicator': 1.0}: 2048]


### Summary - Quantum Variables and Functions Guidelines

The following summarizes the main takeaways from the example in terms of quantum variables and functions:
* There are 3 types of quantum objects in Qmod: a qubit, a qubit array, and a quantum number. Objects can be cast from one to another (as x being cast from quantum number in the main function to qubit array in the flip_msb function)

* Quantum objects are represented by quantum variables. Each variable needs to be declared and initialized before it is used. The initialization can be done in several ways, here we see two options: with the allocate function (the initialization of x) or with a numeric assignment (the initializaiton of indicator).

* Quantum variables that are arguments of a function can be declared with the Output modifier and initialized within the scope of a function (like indicator in the main function). Without the Output modifier they must be initialized outside the scope of the function (like reg in the flip_msb) function.

* Quantum variables can be declared and initialized within a function (like x in the main function)

* The quantum program is always generated from the main function, even when further quantum functions are used. The execution results of the quantum program are interperted only for the arguments of the main function (like for indicator in the above example)

In [6]:
from classiq import write_qmod

write_qmod(create_model(main), "quantum_variables_and_functions")

# Design - Quantum Operations

Quantum computing resembles classical computing in some aspects, and is substantially different in other aspects. One of the key advantages of Qmod is that it captures the core concepts which are uniquely quantum in a simple and natural way. This begins with the quantum objects and variables described in the previous tutorial (see Quantum Variables and Functions) and continues with the meaningful native quantum operations.

Simply put, quantum operations are functions of functions applied on quantum objects that are very common in quantum computing, hence receive a special place in the Qmod language. More accurately, these are built-in statements in the Qmod language. A statement is a building block of a programming language, such that programming languages are composed from statements. For example, if and for loops are common statements in programming languages like in Python.

There are a few quantum operation statements in Qmod, and in the following we focus on arguably the most useful one - control. It applies a specified quantum function conditioned on some state (value) of a given quantum variable. Other quantum operations are invert, power, and within_apply, and one can see all the quantum operations [here](https://docs.classiq.io/latest/reference-manual/platform/qmod/language-reference/statements/control/).

Let's understand it through a concrete example.

### Concrete Example

Our task is to first prepare a quantum number x in a superposition of all possible integers between 0 to 15, i.e. $\Ket{x}=\frac{1}{\sqrt{16}}(\Ket{0}+\Ket{1}+ ... + \Ket{15})$
Then to prepare another quantum number y that is in the state $\Ket{17}$ only if $\Ket{x}$ is in the state $\Ket{15}$, otherwise the state of y should be $\Ket{0}$. Mathematically, the task is to prepare the following state:
$$\Ket{x}\Ket{y}=\frac{1}{\sqrt{16}}(\Ket{0}+\Ket{1}+ ... + \Ket{14}\Ket{0} + \Ket{15}\Ket{17})$$
How to approach this task? We have already seen that we can create a uniform superposition with the `hadamard_transform`. Then, conditioned on the value of $\Ket{x}$ being $\Ket{15}$, we will prepare the state of $\Ket{y}$ to be $\Ket{17}$ with the `inplace_prepare_int` function.


What's the difference between `inplace_prepare_int` and `prepare_int`?
The function `prepare_int` initializes a specific quantum variable to a specific integer value. On the other hand, the function `inplace_prepare_int` receives an initialized quantum variable, and assumed its value is $0$, it prepares its state to the desired integer values. Note that for all the `prepare` functions there is the option to use also `inplace_prepare` in case you want to apply the function on an already initialized quantum variable.

In [7]:
from classiq import QNum, control, inplace_prepare_int, qfunc

@qfunc
def apply_control(x: QNum, y: QNum):
    control(ctrl=(x == 15), operand=lambda: inplace_prepare_int(17, y))

Now for the main function. At the end, we want to evaluate the execution results of both x and y so both of them will be arguments of the main function. x is initialized with 4 qubits since it needs to be in a superposition of $16 = 2^4$ states, and y is initialized with 5 qubits since this is the minimal number of qubits that is needed for representing the number 17 ($floor(log_2(17) = 5)$).

In [8]:
from classiq import Output, allocate, hadamard_transform


@qfunc
def main(x: Output[QNum], y: Output[QNum]):
    allocate(4, x)
    allocate(5, y)

    hadamard_transform(x)
    apply_control(x, y)

In [9]:
from classiq import create_model, show, synthesize

quantum_program = synthesize(create_model(main))
show(quantum_program)

Opening: https://platform.classiq.io/circuit/a547ffc3-7f45-41eb-821e-f1fdb8e313d2?version=0.42.2


Control operation vs quantum gate:
![Circui](./images/Design_1.png)

In [10]:
from classiq import execute

job = execute(quantum_program)
results = job.result()[0].value.parsed_counts
print(results)

[{'x': 5.0, 'y': 0.0}: 147, {'x': 12.0, 'y': 0.0}: 143, {'x': 13.0, 'y': 0.0}: 139, {'x': 6.0, 'y': 0.0}: 139, {'x': 3.0, 'y': 0.0}: 138, {'x': 0.0, 'y': 0.0}: 137, {'x': 10.0, 'y': 0.0}: 133, {'x': 9.0, 'y': 0.0}: 129, {'x': 14.0, 'y': 0.0}: 129, {'x': 7.0, 'y': 0.0}: 128, {'x': 15.0, 'y': 17.0}: 126, {'x': 4.0, 'y': 0.0}: 125, {'x': 2.0, 'y': 0.0}: 115, {'x': 1.0, 'y': 0.0}: 111, {'x': 11.0, 'y': 0.0}: 107, {'x': 8.0, 'y': 0.0}: 102]


We can see that we receive 16 possible values for x and y, and in all the pairs of values $y=0$ besides when $x=15$ as desired!

### Summary - Quantum Operations
* Quantum operations receive both quantum objects and quantum functions as inputs, and they apply the quantum functions on the quantum objects according to the nature of the quantum operation. In the above example, the control operation applies the inplace_prepare_int function on the quantum variable y conditioned on the value of the quantum variable x being.
* In the Native syntax, the quantum operations are statements of the Qmod language and have a special syntax. In the above example, the syntax for control is control(){;} where the condition is given in the parentheses and the operation/function that is applied is within the curly brackets.
* In the Python SDK, the quantum operations are written just like any other Python function. The arguments of the quantum operation that are functions by themselves must be passed with the lambda: keyword. In the above example, the operand argument of the control function is a function by itself (inplace_prepare_int), hence it is prefixed with lambda:.

* Other quantum operations are: power - raising a unitary to some power, invert - applying the inverse of a unitary, within_apply - applying two unitaries U and V in the following way: $UVV^{\dagger}$. See more detailed description of all the quantum operators [here](https://docs.classiq.io/latest/reference-manual/platform/qmod/language-reference/statements/within-apply/).

In [11]:
from classiq import write_qmod

write_qmod(create_model(main), "quantum_operations")

# Design - Classical Variables and Operations

Quantum variables and operations lie at the heart of the Qmod language but they are not enough for designing any quantum algorithm one has in mind. To complete the picture we need also the ability to perform classical control flow with classical operations over classical variables. These are the operations that a typical Python user would expect to have like for loops and if-else statements over int or float variables.
In Qmod, there are equivalent types for the common Python types int, float and bool and they are denoted by int, real and bool in the native syntax and by CInt, CReal and CBool respectively in the Python SDK. In addition, arrays of classical values are also supported as well as more configurable data structures called Structs (which are the equivalent of Python dataclasses without methods). The full list of all the classical variables and types supported in Qmod can be found [here](https://docs.classiq.io/latest/reference-manual/platform/qmod/language-reference/classical-types/#structs).
For now, the only two classical operations supported in Qmod are repeat which is the equivalent of a Python for loop and if-else statements. The detailed description of both can be found [here](https://docs.classiq.io/latest/reference-manual/platform/qmod/language-reference/statements/classical-control-flow/).

Let's understand how to manage classical control flow in Qmod for our benefit through the following concrete example.

### Concrete Example

Our task is to create a quantum array with 10 qubits in the state of $\Ket{1010101010}$, i.e. a quantum state with alternating qubit states of zeros and ones.
How shell we approach this? Knowing that a general qubit array can be easily initialized to the state of all zeros i.e. $\Ket{0000000000}$, we can see that we need to flip the state of every second qubit, i.e. of all the qubit in an even position (the qubit in the 0th position, 2nd position, 4th position etc.)

So we will approach this task in two steps. First we will create a function called apply_condition that receives a qubit and its position within the qubit array, and if the position is even then it flips its state. Secondly, we will go over all the qubits in the qubit array and apply the function on each qubit.

The apply_condition function receives a qubit variable named qubit and a classical integer variable index, and if the condition that the variable index is an even number, then there is a X gate applied on the qubit which flips its state.

In [18]:
from classiq import *

@qfunc
def apply_condition(index: CInt, qubit: QBit):
    if_(condition=index % 2 == 0, then=lambda: X(qubit))

The condition of wether the variable index is even is evaluated using the modulo operation %, that calculates the reminder of the integer index divided by
. One can notice that in the Python SDK there is a use of the lambda: construct before the operation X(qubit). As described in the Quantum Variables and Functions Tutorial, in the Python SDK every function that is given as an argument for another function needs to have the prefix of lambda:.
The argument of the main function is a qubit array named x, and it is initalized to the state $\Ket{0000000000}$ with 10 qubits using the allocate function. Then the classical operator repeat is applied, where function apply_condition is applied 10 times, one time per qubit in the qubit array x.

In [14]:
@qfunc
def main(x: Output[QArray]):
    allocate(10, x)
    repeat(count=x.len, iteration=lambda index: apply_condition(index, x[index]))

The repeat operator has two arguments. First, the number of repetitions is specified, and in this case and it set to the length of the qubit array x. Secondly, the statement (or statements) that are being repeated are specified, and in the above it is the apply_condition function with the two arguments index and the corresponding qubit x[index]. One may notice the different syntax between the Native syntax and the Python SDK in declaring the counting index with the variable index.

In [15]:
quantum_program = synthesize(create_model(main))
show(quantum_program)

Opening: https://platform.classiq.io/circuit/8d545ea1-d1c9-4b86-9152-d327575605c9?version=0.42.2
Opening: https://platform.classiq.io/circuit/91d4f980-1aa3-428f-bab4-4d64de1477cd?version=0.42.2


In [16]:
results = execute(quantum_program).result()[0].value
print(results.counts)

{'0101010101': 2048}


The only measured bit string is 0101010101, which with the Classiq notation should be read from right to left when interperted, hence it corresponds to the state
 $\Ket{1010101010}$ which is exactly what we have tried to achieve!

### Summary - Classical Variables and Operations

The following summarizes the above example in terms of classical variables and operations:
* Qmod supports the use of classical variables. In the above example, the integer variable index is explicitly declared and used. The length of the qubit array x is also a classical integer variable that is used implicitly within the repeat operation.
* Qmod supports the use of classical operations that receive classical variables and quantum operations that are applied according to the classical operation logic. In the above example, the if classical operation applies the quantum operation X(qubit) if the classical variable index is even, and the classical operation repeat applies the quantum operation apply_condition 10 times - the result of evaluating x.len.

In [17]:
write_qmod(create_model(main), "classical_variables_and_operations")